### PR TITLE
MCR-1592 added mods:relatedItems to topological load

### PR DESCRIPTION
**Pull request by KelKx**

### CODE REVIEW

**Comment by MyCoRe-Travis** [comment]
Something gone wrong on 210669faa83e56f32c323ce2b98677a4620cb11d! 
 Please Check https://github.com/MyCoRe-Travis/test_artifacts/tree/1324-pull_request-235

**Comment by MyCoRe-Travis** [comment]
Something gone wrong on 210669faa83e56f32c323ce2b98677a4620cb11d! 
 Please Check https://github.com/MyCoRe-Travis/test_artifacts/tree/1324-pull_request-235

**Comment by MyCoRe-Travis** [comment]
Something gone wrong on 210669faa83e56f32c323ce2b98677a4620cb11d! 
 Please Check https://github.com/MyCoRe-Travis/test_artifacts/tree/1324-pull_request-235

File: mycore-base/src/main/java/org/mycore/tools/MCRTopologicalSort.java [hunk 1]
```diff
@@ -104,8 +105,25 @@ public void prepareData(String[] files, File dir) {
                             if (xmlStreamReader.getLocalName().equals("mycoreobject")) {
                                 nodes.forcePut(i, xmlStreamReader.getAttributeValue(null, "ID"));
                             } else if (xmlStreamReader.getLocalName().equals("parent")) {
-                                parentNames.put(i,
+                                List<String> dependecyList;
```
Comment (yagee-de) [inline mycore-base/src/main/java/org/mycore/tools/MCRTopologicalSort.java]: spelling

File: mycore-base/src/main/java/org/mycore/tools/MCRTopologicalSort.java [hunk 2]
```diff
@@ -104,8 +105,25 @@ public void prepareData(String[] files, File dir) {
...
+                            } else if (xmlStreamReader.getLocalName().equals("relatedItem")){
+                                List<String> dependecyList;
+                                if(parentNames.containsKey(i)){
+                                    dependecyList = parentNames.get(i);
+                                } else {
+                                    dependecyList= new ArrayList<>();
+                                    parentNames.put(i, dependecyList);
+                                }
+                                dependecyList.add(
+                                    xmlStreamReader.getAttributeValue("http://www.w3.org/1999/xlink", "href" ));
```
Comment (yagee-de) [inline mycore-base/src/main/java/org/mycore/tools/MCRTopologicalSort.java]: check that href is a MCRObjectID

File: mycore-base/src/main/java/org/mycore/tools/MCRTopologicalSort.java [hunk 3]
```diff
@@ -127,10 +147,14 @@ public void prepareData(String[] files, File dir) {
...
         for (int source : parentNames.keySet()) {
-            Integer target = nodes.inverse().get(parentNames.get(source));
-            if (target != null) {
-                addEdge(source, target);
+            List<String> dependencies = parentNames.get(source);
+            for(String element: dependencies){
+                Integer target = nodes.inverse().get(element);
+                if (target != null) {
+                    addEdge(source, target);
+                }
```
Comment (yagee-de) [inline mycore-base/src/main/java/org/mycore/tools/MCRTopologicalSort.java]: Perfect example for using Streams:
```
parentNames.get(source)
           .stream()
           .map(nodes.inverse()::get)
           .filter(Objects::nonNull)
           .forEach(target -> addEdge(source, target))
```

File: mycore-base/src/main/java/org/mycore/tools/MCRTopologicalSort.java [hunk 4]
```diff
@@ -104,8 +105,25 @@ public void prepareData(String[] files, File dir) {
                             if (xmlStreamReader.getLocalName().equals("mycoreobject")) {
                                 nodes.forcePut(i, xmlStreamReader.getAttributeValue(null, "ID"));
                             } else if (xmlStreamReader.getLocalName().equals("parent")) {
-                                parentNames.put(i,
+                                List<String> dependecyList;
+                                if(parentNames.containsKey(i)){
```
Comment (yagee-de) [inline mycore-base/src/main/java/org/mycore/tools/MCRTopologicalSort.java]: `dependencyList = parentNames.computeIfAbsent(i, i -> new ArrayList())` is much shorter than using the complicated if-then-else

File: mycore-base/src/main/java/org/mycore/tools/MCRTopologicalSort.java [hunk 5]
```diff
@@ -104,8 +105,25 @@ public void prepareData(String[] files, File dir) {
...
+                                    dependecyList = parentNames.get(i);
+                                } else {
+                                     dependecyList= new ArrayList<>();
+                                     parentNames.put(i, dependecyList);
+                                }
+                               dependecyList.add(
                                     xmlStreamReader.getAttributeValue("http://www.w3.org/1999/xlink", "href"));
+                            } else if (xmlStreamReader.getLocalName().equals("relatedItem")){
+                                List<String> dependecyList;
```
Comment (yagee-de) [inline mycore-base/src/main/java/org/mycore/tools/MCRTopologicalSort.java]: same here: use `java.util.Map.computeIfAbsent(K, Function<? super K, ? extends V>)`

**Review by yagee-de — CHANGES_REQUESTED** [review]

**Review by yagee-de — APPROVED** [review]

**Pull request merged**